File: day12/solver.py
import sys

sys.path.insert(0, "../utils/py")
import utils
import rich
from queue import Queue


measure_time = utils.stopwatch()


@measure_time
def parse(raw_data):
    graph = {}
    for y, line in enumerate(raw_data.splitlines()):
        for x, c in enumerate(line):
            graph[(x, y)] = c
    start = [p for p in graph if graph[p] == "S"][0]
    end   = [p for p in graph if graph[p] == "E"][0]
    graph[start] = "a"
    graph[end] = "z"
    return graph, start, end

def neighbors(graph, p):
    x, y = p
    adjacent = [(x-1,y), (x+1,y), (x,y-1), (x,y+1)]
    return [n for n in adjacent if (n[0]>=0 and n[1]>=0) and n in graph and ord(graph[n]) <= ord(graph[p])+1]

def breadthfirstsearch(graph, start, end):
    queue = Queue()
    queue.put(start)
    path = {start: None}
    while not queue.empty():
        v = queue.get()
        if v == end:
            return path
        for n in neighbors(graph, v):
            if n not in path:
                path[n] = v
                queue.put(n)
    return path


def shortestpath(graph, start, end):
    pathmap = breadthfirstsearch(graph, start, end)
    if end not in pathmap:
        return []
    path = [end]
    while parent:=pathmap[path[-1]]:
        path.append(parent)
    return list(reversed(path))

# PART 1
@measure_time
def solve1(data):
    return len(shortestpath(*data)) - 1


# PART 2
@measure_time
def solve2(data):
    graph, start, end = data
    starts = [p for p in graph if graph[p] == "a"]
    lengths = []
    for s in starts:
        # this is stupid
        path = shortestpath(graph, s, end)
        if path:
            lengths.append(len(path) - 1)
    return min(lengths)


if __name__ == "__main__":
    data = parse(open("input.txt").read().strip())
    print("Part 1: {}".format(solve1(data)))
    print("Part 2: {}".format(solve2(data)))

    print("\nTime taken:")
    for func, time in measure_time.times:
        print(f"{func:8}{time}s")
    print("----------------")
    print("total   {}s".format(sum(t for _, t in measure_time.times)))

File: day12/solver.hs
{-# LANGUAGE QuasiQuotes #-}

import Text.RawString.QQ 
import Data.List
import Utils
import AoC
import Geometry
import Data.Vector (Vector)
import Data.Maybe (fromJust, isNothing)
import qualified Data.Map as Map
import Data.Map (Map)
import Data.Char (ord)

getInput :: IO String
getInput = readFile "input.txt"

--------------------------------------------------------------------

type Parsed = Grid Int Char
type Node = (Vector Int, Char)
type Sol1 = Int
type Sol2 = Int

parse :: String -> Parsed
parse = gridFromString

neighbors :: Parsed -> Node -> [Node]
neighbors grid (x,a) = [(xn, grid Map.! xn)
    | xn <- [left x, right x, up x, down x],
    xn `Map.member` grid, ord (grid Map.! xn) <= (ord a) + 1
    ]

breadthFirstSearch :: Parsed -> Node -> Node -> Map Node (Maybe Node)
breadthFirstSearch graph start end = bfs (Map.singleton start Nothing) [start] end
    where
        bfs path [] end = path
        bfs path (q:queue) end
            | q == end = path
            | otherwise = bfs (Map.union newparents path) (ns ++ queue) end
            where
                ns = [n | n <- neighbors graph q, not (n `Map.member` path)]
                newparents = Map.fromList $ zip ns (repeat (Just q))

shortestFromBFS :: Node -> Node -> Map Node (Maybe Node) -> [Node]
shortestFromBFS start end parents
    | not $ end `Map.member` parents = []
    | otherwise = shortest' [] parents start end
    where
        shortest' path parents start end = case parents Map.! end of
            Nothing -> path
            Just p -> shortest' (p:path) parents start p


solve1 :: Parsed -> Sol1
solve1 parsed = length . shortestFromBFS start end $ breadthFirstSearch grid start end
    where
        start = (fromJust $ lookupValue 'S' parsed, 'a')
        end = (fromJust $ lookupValue 'E' parsed, 'z')
        grid = uncurry Map.insert start $ uncurry Map.insert end parsed

solve2 :: Parsed -> Sol2
solve2 parsed = undefined


testdata = [r|Sabqponm
abcryxxl
accszExk
acctuvwj
abdefghi|]
testresult1 = 31
testresult2 = 0

--------------------------------------------------------------------

test1 = test (solve1 . parse) testresult1 testdata
test2 = test (solve2 . parse) testresult2 testdata

printPart :: (Show a) => Int -> a -> IO ()
printPart part solution = do
    putStr $ "Part " ++ (show part) ++ ": "
    print solution

main' :: CLIOptions -> IO ()
main' (CLIOptions p False) = do
    parsed <- parse <$> getInput
    case p of
        1 -> printPart 1 (solve1 parsed)
        2 -> printPart 2 (solve2 parsed)
        _ -> (do
            printPart 1 (solve1 parsed)
            printPart 2 (solve2 parsed))
main' (CLIOptions p True) = 
    case p of
        1 -> printPart 1 test1
        2 -> printPart 2 test2
        _ -> (do
            printPart 1 test1
            printPart 2 test2)

main = main' =<< clioptions
